File: backend/src/agents/agent.py
from __future__ import annotations

import asyncio
import json
import re
from typing import Any, AsyncGenerator, Optional

from src.agents.providers.registry import ProviderRegistry
from src.agents.sandbox.registry import SandboxRegistry
from src.agents.subagents import SubAgentRunner
from src.agents.systemprompts.systemprompt import SYSTEM_PROMPT
from src.agents.tools.registry import ToolRegistry
from src.schemas.chat import ChatMessage, ChatStreamRequest
from src.services.event_buffer import SessionEventBuffer
from src.services.session_store import SessionStore


class AgentRunner:
    def __init__(
        self,
        *,
        provider_registry: ProviderRegistry,
        sandbox_registry: SandboxRegistry,
        tool_registry: ToolRegistry,
        session_store: SessionStore,
        sub_agent_runner: SubAgentRunner,
    ) -> None:
        self.provider_registry = provider_registry
        self.sandbox_registry = sandbox_registry
        self.tool_registry = tool_registry
        self.session_store = session_store
        self.sub_agent_runner = sub_agent_runner

    async def run_agent(self, request: ChatStreamRequest, buffer: SessionEventBuffer) -> None:
        async def send(event: str, data: dict) -> None:
            await buffer.append(event, data)

        try:
            try:
                session = self.session_store.upsert_history(request.chat_id, request.history)
                session.messages.append(ChatMessage(role="user", content=request.user_message).model_dump(exclude_none=True))
            except Exception as exc:
                await send("error", {"message": f"Session setup failed: {exc}", "code": "session_setup_failed"})
                await send("done", {"ok": False})
                return

            await send("iteration", {"current": 0, "limit": request.max_iterations})

            try:
                sandbox_adapter = self.sandbox_registry.get(request.sandbox.provider)
            except Exception as exc:
                await send("error", {"message": f"Sandbox adapter error: {exc}", "code": "sandbox_adapter_error"})
                await send("done", {"ok": False})
                return

            if session.sandbox_context is None:
                await send("status", {"state": "creating_sandbox", "label": "Creating sandbox..."})
                try:
                    session.sandbox_context = await sandbox_adapter.create(request.sandbox)
                except Exception as exc:
                    await send("error", {"message": f"Sandbox creation failed: {exc}", "code": "sandbox_create_failed"})
                    await send("done", {"ok": False})
                    return
                await send(
                    "sandbox",
                    {
                        "sandbox_id": session.sandbox_context.sandbox_id,
                        "provider": session.sandbox_context.provider,
                        "root_path": session.sandbox_context.root_path,
                    },
                )

            try:
                provider = self.provider_registry.get(request.provider)
            except Exception as exc:
                await send("error", {"message": f"Provider error: {exc}", "code": "provider_error"})
                await send("done", {"ok": False})
                return

            visible_answer_parts: list[str] = []
            visible_reasoning_parts: list[str] = []
            iteration = 0

            while iteration < request.max_iterations:
                iteration += 1
                await send("iteration", {"current": iteration, "limit": request.max_iterations})
                await send("status", {"state": "thinking", "label": "Thinking..."})

                assistant_content_parts: list[str] = []
                assistant_reasoning_parts: list[str] = []
                streamed_tool_calls: list[dict] = []
                finish_reason: Optional[str] = None

                try:
                    async for delta in provider.stream_chat_completion(
                        api_key=request.api_key,
                        model=request.model,
                        messages=self._build_provider_messages(session.messages),
                        tools=self.tool_registry.schemas,
                        base_url=request.base_url,
                    ):
                        if delta.text:
                            cleaned = self._normalize_delta(delta.text)
                            if cleaned:
                                assistant_content_parts.append(cleaned)
                                visible_answer_parts.append(cleaned)
                                await send("token", {"value": cleaned})
                        if delta.reasoning:
                            cleaned = self._normalize_delta(delta.reasoning)
                            if cleaned:
                                assistant_reasoning_parts.append(cleaned)
                                visible_reasoning_parts.append(cleaned)
                                await send("reasoning", {"value": cleaned})
                        if delta.tool_calls:
                            streamed_tool_calls = self._merge_tool_calls(streamed_tool_calls, delta.tool_calls)
                        if delta.finish_reason:
                            finish_reason = delta.finish_reason
                except Exception as exc:
                    await send(
                        "error",
                        {
                            "message": f"Provider API error: {exc}",
                            "code": "provider_api_error",
                        },
                    )
                    await send("done", {"ok": False})
                    return

                try:
                    if streamed_tool_calls or finish_reason == "tool_calls":
                        assistant_tool_message = {
                            "role": "assistant",
                            "content": "".join(assistant_content_parts) or None,
                            "tool_calls": streamed_tool_calls,
                        }
                        if assistant_reasoning_parts:
                            assistant_tool_message["reasoning_content"] = "".join(assistant_reasoning_parts)
                        session.messages.append(assistant_tool_message)

                        for tool_call in streamed_tool_calls:
                            tool_name = tool_call.get("function", {}).get("name", "unknown")
                            tool_args = tool_call.get("function", {}).get("arguments", "{}")
                            tool_payload = self._safe_json_loads(tool_args)
                            file_path = tool_payload.get("file_path")
                            command = tool_payload.get("command")
                            session_name = tool_payload.get("session_name") or tool_payload.get("session") or "default"
                            session_names = tool_payload.get("session_names")
                            list_path = tool_payload.get("path")
                            query = tool_payload.get("query")
                            url = tool_payload.get("url")
                            old_string = tool_payload.get("old_string")
                            new_string = tool_payload.get("new_string")
                            raw_input = tool_payload.get("input")
                            sub_agent = tool_payload.get("agent") if tool_name == "call_sub_agent" else None
                            await send(
                                "tool_call",
                                {
                                    "name": tool_name,
                                    "file_path": file_path,
                                    "command": command,
                                    "session": tool_payload.get("session") if tool_name == "call_sub_agent" else None,
                                    "agent": sub_agent,
                                    "session_name": session_name,
                                    "session_names": session_names,
                                    "path": list_path,
                                    "query": query,
                                    "url": url,
                                    "old_string": old_string,
                                    "new_string": new_string,
                                    "input": raw_input,
                                    "label": self._tool_label(tool_name, file_path, command, list_path, session_names, url=url, agent=sub_agent),
                                },
                            )

                            tool_task = asyncio.create_task(
                                self.tool_registry.execute(
                                    tool_name,
                                    tool_args,
                                    sandbox_adapter=sandbox_adapter,
                                    sandbox_context=session.sandbox_context,
                                    provider=provider,
                                    model=request.model,
                                    api_key=request.api_key,
                                    base_url=request.base_url,
                                    chat_id=request.chat_id,
                                    session_store=self.session_store,
                                    buffer=buffer,
                                    sub_agent_runner=self.sub_agent_runner,
                                    sub_agents=request.sub_agents,
                                    tavily_api_key=request.tavily_api_key,
                                    exa_api_key=request.exa_api_key,
                                    serpapi_api_key=request.serpapi_api_key,
                                    search_provider=request.search_provider,
                                    firecrawl_api_key=request.firecrawl_api_key,
                                )
                            )

                            try:
                                result = await tool_task
                            except Exception as exc:
                                result = {"ok": False, "error": {"code": "tool_execution_failed", "message": str(exc)}}

                            session.messages.append(
                                {
                                    "role": "tool",
                                    "tool_call_id": tool_call.get("id"),
                                    "name": tool_name,
                                    "content": json.dumps(result),
                                    "metadata": {"file_path": file_path},
                                }
                            )

                            await send(
                                "tool_result",
                                {
                                    "name": tool_name,
                                    "file_path": file_path,
                                    "ok": result.get("ok", False),
                                    "result": result,
                                },
                            )
                        continue

                    final_message = "".join(assistant_content_parts)
                    final_assistant_message: dict[str, Any] = {"role": "assistant", "content": final_message}
                    if assistant_reasoning_parts:
                        final_assistant_message["reasoning_content"] = "".join(assistant_reasoning_parts)
                    session.messages.append(final_assistant_message)
                    await send(
                        "message_complete",
                        {
                            "content": "".join(visible_answer_parts),
                            "iteration_count": iteration,
                            "reasoning": "".join(visible_reasoning_parts) if visible_reasoning_parts else None,
                        },
                    )

                    await send("done", {"ok": True})
                    return

                except Exception as exc:
                    await send(
                        "error",
                        {
                            "message": f"Agent loop error: {exc}",
                            "code": "agent_loop_error",
                        },
                    )
                    await send("done", {"ok": False})
                    return

            await send(
                "error",
                {
                    "message": "Iteration limit reached before the agent could finish the turn.",
                    "code": "iteration_limit_reached",
                },
            )
            await send("done", {"ok": False})
        except asyncio.CancelledError:
            await send("done", {"ok": False})
            raise
        finally:
            buffer.set_done()

    async def stream_sse(self, chat_id: str, since_event_id: int = -1) -> AsyncGenerator[str, None]:
        session = self.session_store.get(chat_id)
        if not session or not session.event_buffer:
            yield self._sse("error", {"message": "No active session or event buffer.", "code": "no_session"})
            yield self._sse("done", {"ok": False})
            return

        async for event_data in session.event_buffer.subscribe(since_event_id):
            yield self._sse(event_data["event"], event_data["data"])

    def _build_provider_messages(self, messages: list[dict]) -> list[dict]:
        built_messages = [{"role": "system", "content": SYSTEM_PROMPT}]
        for message in messages:
            built: dict = {"role": message["role"]}
            if message.get("content") is not None:
                built["content"] = message.get("content")
            if message.get("reasoning_content") is not None:
                built["reasoning_content"] = message.get("reasoning_content")
            if message.get("tool_calls") is not None:
                built["tool_calls"] = message.get("tool_calls")
            if message.get("tool_call_id") is not None:
                built["tool_call_id"] = message.get("tool_call_id")
            if message.get("name") is not None:
                built["name"] = message.get("name")
            built_messages.append(built)
        return built_messages

    def _merge_tool_calls(self, accumulated: list[dict], incoming: list[dict]) -> list[dict]:
        merged = accumulated[:] if accumulated else []
        for chunk in incoming:
            index = chunk.get("index", len(merged))
            while len(merged) <= index:
                merged.append({"id": None, "type": "function", "function": {"name": "", "arguments": ""}})
            target = merged[index]
            if chunk.get("id"):
                target["id"] = chunk["id"]
            if chunk.get("type"):
                target["type"] = chunk["type"]
            function = chunk.get("function") or {}
            target_function = target.setdefault("function", {"name": "", "arguments": ""})
            if function.get("name"):
                target_function["name"] += function["name"]
            if function.get("arguments"):
                target_function["arguments"] += function["arguments"]
        return merged

    def _safe_json_loads(self, raw: str) -> dict:
        try:
            return json.loads(raw or "{}")
        except json.JSONDecodeError:
            match = re.search(r"\{.*\}", raw or "")
            if not match:
                return {}
            try:
                return json.loads(match.group(0))
            except json.JSONDecodeError:
                return {}

    def _tool_label(self, tool_name: str, file_path: Optional[str], command: Optional[str] = None, list_path: Optional[str] = None, session_names: Optional[list[str]] = None, url: Optional[str] = None, **kwargs) -> str:
        if tool_name == "call_sub_agent":
            return f"Sub-Agent: {kwargs.get('agent', 'unknown')}"
        if tool_name == "list_sub_agents":
            return "List Sub-Agents"
        if tool_name == "shall_tool":
            return f"Terminal: {command or 'unknown'}"
        if tool_name == "shell_view":
            sessions = ", ".join(session_names) if session_names else "unknown"
            return f"Shell View: {sessions}"
        if tool_name == "list_files":
            return f"List: {list_path or 'unknown'}"
        if tool_name == "web_search":
            return f"Web Search"
        if tool_name == "fatch_web_urls":
            return f"Fetch: {url or 'url'}"
        if tool_name == "apply_patch":
            return f"Apply Patch"
        if tool_name == "str_replace":
            return f"Edit: {file_path or 'unknown'}"
        prefix = "Create" if tool_name == "file_write" else "Read"
        return f"{prefix}: {file_path or 'unknown'}"

    def _normalize_delta(self, text: str) -> str:
        return text.replace("\r\n", "\n")

    def _sse(self, event: str, data: dict) -> str:
        return f"event: {event}\ndata: {json.dumps(data)}\n\n"

File: curro-ai/src/agents/agent.ts
import type { AppConfig } from "../config.js";
import { buildSystemPrompt } from "./systemprompt.js";
import type { ProviderRegistry } from "./providers/registry.js";
import { resolveProvider } from "./providers/registry.js";
import type { Provider } from "./providers/types.js";
import type { ToolRegistry } from "./tools/index.js";
import {
  buildImageMessage,
  extractImageAttachment,
  withoutImageAttachment,
} from "./tools/readImage.js";
import { isVisionCapableModel } from "../utils/vision.js";
import type { SkillDefinition, SubAgentDefinition, WebToolsConfig } from "./tools/types.js";
import type { ToolCall, ToolCallDelta } from "./providers/types.js";
import type { ChatSession, StoredMessage } from "../services/sessionStore.js";
import type { SessionEventBuffer } from "../services/eventBuffer.js";
import type { PlanApprovalStore } from "../services/planApprovalStore.js";
import type { QuestionStore } from "../services/questionStore.js";
import { safeJsonParse } from "../utils/json.js";
import { createSubAgentRuntime } from "./subagents.js";
import { createSkillRuntime } from "./skills.js";
import { mergeDefaultSkills, resolveDefaultSkills } from "./skills/index.js";
import { resolveDefaultSubAgents, mergeDefaultSubAgents } from "./sub-agents/index.js";
import { createTodoRuntime } from "./todos.js";
import { createMemoryRuntime } from "./memory.js";
import { createKnowledgeRuntime } from "./knowledge.js";
import type { KnowledgeFile, MemoryFile, TodoItem } from "./tools/types.js";

export interface RunAgentRequest {
  chatId: string;
  userMessage: string;
  provider: string;
  model: string;
  apiKey: string;
  baseUrl?: string;
  /** Full user-defined provider config, present when a `custom_` provider is selected. */
  customProvider?: unknown;
  maxIterations: number;
  temperature?: number;
  /** Per-request web tool keys/provider (from frontend Settings); falls back to env config. */
  tavilyApiKey?: string;
  exaApiKey?: string;
  serpapiApiKey?: string;
  searchProvider?: "duckduckgo" | "tavily" | "exa" | "serpapi";
  fetchProvider?: "builtin" | "firecrawl";
  firecrawlApiKey?: string;
  /** User-defined sub-agents (from the frontend, stored in the browser) available this turn. */
  subAgents?: SubAgentDefinition[];
  /** User-defined skills (from the frontend, stored in the browser) available this turn. */
  skills?: SkillDefinition[];
  /** User-defined todos (from the frontend, stored in the browser) available this turn. */
  todos?: TodoItem[];
  /** The user's memory files (from the frontend, stored in the browser) available this turn. */
  memory?: MemoryFile[];
  /** The user's knowledge base files (from the frontend, stored in the browser) available this turn. */
  knowledge?: KnowledgeFile[];
}

export class AgentRunner {
  constructor(
    private readonly providers: ProviderRegistry,
    private readonly tools: ToolRegistry,
    private readonly config: AppConfig,
    private readonly planApprovals: PlanApprovalStore,
    private readonly askQuestions: QuestionStore,
  ) {}

  /**
   * Execute a full autonomous turn: stream reasoning + tokens, run tools natively, and
   * loop (Thought -> Action -> Observation) until the model produces a final answer with
   * no further tool calls or the iteration limit is hit.
   */
  async run(
    request: RunAgentRequest,
    session: ChatSession,
    buffer: SessionEventBuffer,
    signal: AbortSignal,
  ): Promise<void> {
    const send = (event: string, data: Record<string, unknown>) => buffer.append(event, data);

    try {
      let provider: Provider;
      try {
        provider = resolveProvider(this.providers, request.provider, request.customProvider);
      } catch (error) {
        send("error", { code: "provider_error", message: messageOf(error) });
        send("done", { ok: false });
        return;
      }

      // Memory runtime for this turn — a browser-backed snapshot of the user's memory files that
      // the `memory` tool reads and mutates. Present only for main-agent tool calls. The three
      // pre-added files (MEMORY.md, SOUL.md, USER.md) are auto-appended to the user's FIRST message
      // of a chat so the agent starts every conversation already knowing its accumulated context.
      const memoryRuntime = createMemoryRuntime(request.memory ?? []);

      // Knowledge runtime for this turn — a browser-backed snapshot of the user's knowledge base
      // (a file-tree of durable reference material) that the knowledge_* tools read and mutate.
      // Present only for main-agent tool calls. Unlike memory it has no pre-added files; a one-time
      // notice is appended to the user's FIRST message ONLY when the user actually has knowledge
      // files, telling the agent to discover/read them with the knowledge tools.
      const knowledgeRuntime = createKnowledgeRuntime(request.knowledge ?? []);

      // "First user input" = no prior user turn exists in this session's history. On that first
      // message (and only then) we prepend the pre-added memory files so the model has its
      // persistent context; later turns rely on the `memory` tool to read anything it needs. The
      // knowledge-base notice is likewise appended only on this first message (and only if any
      // knowledge files exist).
      const isFirstUserMessage = !session.messages.some((m) => m.role === "user");
      const userContent = isFirstUserMessage
        ? withFirstMessageContext(
            request.userMessage,
            memoryRuntime.firstMessageContext(),
            knowledgeRuntime.firstMessageContext(),
          )
        : request.userMessage;

      session.messages.push({ role: "user", content: userContent });
      const systemPrompt = buildSystemPrompt(this.config.workspaceRoot);
      const maxIterations = clampIterations(request.maxIterations, this.config.maxIterations);
      const visionCapable = isVisionCapableModel(request.model, this.config);
      const web: WebToolsConfig = {
        searchProvider: request.searchProvider ?? this.config.searchProvider,
        fetchProvider: request.fetchProvider ?? this.config.fetchProvider,
        tavilyApiKey: request.tavilyApiKey || this.config.tavilyApiKey || undefined,
        exaApiKey: request.exaApiKey || this.config.exaApiKey || undefined,
        serpapiApiKey: request.serpapiApiKey || this.config.serpapiApiKey || undefined,
        firecrawlApiKey: request.firecrawlApiKey || this.config.firecrawlApiKey || undefined,
      };

      // Sub-agent runtime for this turn — a separate LLM call per sub-agent, using the same
      // provider/model/credentials the user chose for the main agent. Streams `sub_agent_*`
      // side-channel events onto the same buffer so the UI can show live progress. The agent's
      // built-in default sub-agents are merged underneath so they are pre-added and always
      // available, unless the user provides their own sub-agent with the same name (which
      // overrides the default).
      const defaultSubAgents = await resolveDefaultSubAgents();
      const subAgentRuntime = createSubAgentRuntime({
        provider,
        tools: this.tools,
        config: this.config,
        chatId: request.chatId,
        definitions: mergeDefaultSubAgents(defaultSubAgents, request.subAgents ?? []),
        model: request.model,
        apiKey: request.apiKey,
        baseUrl: request.baseUrl,
        temperature: request.temperature,
        send,
        // Snapshot the live conversation so a call_sub_agent with send_my_context=true can share
        // the surrounding context. Returns the current messages at the moment the sub-agent is called.
        getConversationContext: () => session.messages,
      });

      // Skill runtime for this turn — enumerates the user's skills (list_skills) and materializes
      // them onto disk inside the workspace's ".curro/skills" directory (skill_initialize). Skills, like
      // sub-agents, are authored in the frontend and travel with each turn. The agent's built-in
      // default skills are merged underneath so they are pre-added and always available, unless
      // the user provides their own skill with the same name (which overrides the default).
      const defaultSkills = await resolveDefaultSkills();
      const skillRuntime = createSkillRuntime(mergeDefaultSkills(defaultSkills, request.skills ?? []));

      // Todo runtime for this turn — a browser-backed snapshot of the user's todo list that the
      // TodoWrite / read_todos tools read and update. Present only for main-agent tool calls.
      const todoRuntime = createTodoRuntime(request.todos ?? []);

      const visibleAnswer: string[] = [];
      const visibleReasoning: string[] = [];
      let iteration = 0;

      send("iteration", { current: 0, limit: maxIterations });

      while (iteration < maxIterations) {
        if (signal.aborted) {
          send("done", { ok: false, aborted: true });
          return;
        }

        iteration += 1;
        send("iteration", { current: iteration, limit: maxIterations });
        send("status", { state: "thinking", label: "Thinking..." });

        const answerParts: string[] = [];
        const reasoningParts: string[] = [];
        let toolCalls: ToolCall[] = [];
        let finishReason: string | null = null;

        try {
          const stream = provider.streamChatCompletion({
            apiKey: request.apiKey,
            model: request.model,
            messages: this.buildProviderMessages(systemPrompt, session.messages),
            tools: this.tools.schemas,
            baseUrl: request.baseUrl,
            temperature: request.temperature,
            signal,
          });

          for await (const delta of stream) {
            if (delta.reasoning) {
              const cleaned = normalize(delta.reasoning);
              reasoningParts.push(cleaned);
              visibleReasoning.push(cleaned);
              send("reasoning", { value: cleaned });
            }
            if (delta.text) {
              const cleaned = normalize(delta.text);
              answerParts.push(cleaned);
              visibleAnswer.push(cleaned);
              send("token", { value: cleaned });
            }
            if (delta.toolCalls) {
              toolCalls = mergeToolCalls(toolCalls, delta.toolCalls);
            }
            if (delta.finishReason) {
              finishReason = delta.finishReason;
            }
          }
        } catch (error) {
          if (signal.aborted) {
            send("done", { ok: false, aborted: true });
            return;
          }
          send("error", { code: "provider_api_error", message: `Provider API error: ${messageOf(error)}` });
          send("done", { ok: false });
          return;
        }

        const hasToolCalls = toolCalls.some((c) => c.function.name) || finishReason === "tool_calls";

        if (hasToolCalls) {
          const assistantMessage: StoredMessage = {
            role: "assistant",
            content: answerParts.join("") || null,
            tool_calls: toolCalls.map((c) => ({
              id: c.id,
              type: c.type,
              function: { name: c.function.name, arguments: c.function.arguments },
            })),
          };
          if (reasoningParts.length > 0) assistantMessage.reasoning_content = reasoningParts.join("");
          session.messages.push(assistantMessage);

          const imageMessages: StoredMessage[] = [];

          for (const toolCall of toolCalls) {
            if (!toolCall.function.name) continue;
            const args = safeJsonParse(toolCall.function.arguments);

            send("tool_call", {
              id: toolCall.id,
              name: toolCall.function.name,
              args,
              label: this.tools.label(toolCall.function.name, args),
            });

            const result = await this.tools.execute(toolCall.function.name, args, {
              workspaceRoot: this.config.workspaceRoot,
              shellTimeoutMs: this.config.shellTimeoutMs,
              signal,
              web,
              subAgents: subAgentRuntime,
              skills: skillRuntime,
              todos: todoRuntime,
              memory: memoryRuntime,
              knowledge: knowledgeRuntime,
              toolCallId: toolCall.id ?? undefined,
              chatId: request.chatId,
              emit: send,
              planApprovals: this.planApprovals,
              planApprovalTimeoutMs: this.config.planApprovalTimeoutMs,
              askQuestions: this.askQuestions,
              questionTimeoutMs: this.config.questionTimeoutMs,
              model: request.model,
              visionCapable,
              availableToolNames: this.tools.names(),
            });

            // read_image attaches the loaded image to its result. The base64 payload
            // is stripped from the model-visible tool message (it is useless as text
            // and would bloat every follow-up request); the image is instead injected
            // as a vision content part after all tool responses so the model can see it.
            const resultForModel = withoutImageAttachment(result);
            const imageAttachment = extractImageAttachment(result);

            session.messages.push({
              role: "tool",
              tool_call_id: toolCall.id ?? undefined,
              name: toolCall.function.name,
              content: JSON.stringify(resultForModel),
            });

            if (imageAttachment) imageMessages.push(buildImageMessage(imageAttachment));

            send("tool_result", {
              id: toolCall.id,
              name: toolCall.function.name,
              ok: result.ok,
              result: resultForModel,
              label: this.tools.label(toolCall.function.name, args),
            });
          }

          // Append any vision inputs after all tool responses so providers see
          // contiguous tool messages immediately following the assistant tool_calls.
          for (const imageMessage of imageMessages) {
            session.messages.push(imageMessage);
          }

          // Observation delivered — loop again so the model can reason about the results.
          continue;
        }

        // No tool calls -> this is the final assistant answer for the turn.
        const finalContent = answerParts.join("");
        const finalMessage: StoredMessage = { role: "assistant", content: finalContent };
        if (reasoningParts.length > 0) finalMessage.reasoning_content = reasoningParts.join("");
        session.messages.push(finalMessage);

        send("message_complete", {
          content: visibleAnswer.join(""),
          reasoning: visibleReasoning.length > 0 ? visibleReasoning.join("") : null,
          iteration_count: iteration,
        });
        send("done", { ok: true });
        return;
      }

      send("error", {
        code: "iteration_limit_reached",
        message: "Reached the maximum number of iterations before completing the turn.",
      });
      send("done", { ok: false });
    } finally {
      buffer.setDone();
      session.running = false;
      session.updatedAt = Date.now();
    }
  }

  private buildProviderMessages(systemPrompt: string, messages: StoredMessage[]): Array<Record<string, unknown>> {
    const built: Array<Record<string, unknown>> = [{ role: "system", content: systemPrompt }];
    for (const message of messages) {
      const entry: Record<string, unknown> = { role: message.role };
      if (message.content !== undefined && message.content !== null) entry.content = message.content;
      else if (message.role === "assistant" && message.tool_calls) entry.content = null;
      if (message.tool_calls) entry.tool_calls = message.tool_calls;
      if (message.tool_call_id) entry.tool_call_id = message.tool_call_id;
      if (message.name) entry.name = message.name;
      // Reasoning-capable models (e.g. OpenCode Zen thinking mode, DeepSeek) require the
      // reasoning_content of a prior assistant message to be passed back unchanged on the
      // next request; omitting it triggers a 400 invalid_request_error.
      if (message.reasoning_content) entry.reasoning_content = message.reasoning_content;
      built.push(entry);
    }
    return built;
  }
}

/**
 * Prepend the one-time context blocks to the user's FIRST message of a chat: the pre-added memory
 * files (MEMORY.md/SOUL.md/USER.md), then the knowledge-base notice (only present when the user has
 * knowledge files). Each block is added only when non-empty; the raw message is returned unchanged
 * when there is nothing to add.
 */
function withFirstMessageContext(
  userMessage: string,
  memoryContext: string,
  knowledgeContext: string,
): string {
  const blocks = [memoryContext.trim(), knowledgeContext.trim()].filter((b) => b.length > 0);
  if (blocks.length === 0) return userMessage;
  return `${blocks.join("\n\n")}\n\n${userMessage}`;
}

function clampIterations(requested: number, max: number): number {
  if (!Number.isFinite(requested) || requested < 1) return max;
  return Math.min(Math.floor(requested), max);
}

function normalize(text: string): string {
  return text.replace(/\r\n/g, "\n");
}

function messageOf(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

/** Merge streamed tool-call fragments into complete tool calls, keyed by index. */
function mergeToolCalls(accumulated: ToolCall[], incoming: ToolCallDelta[]): ToolCall[] {
  const merged = accumulated.slice();
  for (const chunk of incoming) {
    const index = chunk.index ?? merged.length;
    while (merged.length <= index) {
      merged.push({ id: null, type: "function", function: { name: "", arguments: "" } });
    }
    const target = merged[index]!;
    if (chunk.id) target.id = chunk.id;
    if (chunk.function?.name) target.function.name += chunk.function.name;
    if (chunk.function?.arguments) target.function.arguments += chunk.function.arguments;
  }
  return merged;
}
